### PR TITLE
Unconfirmed order line manipulation

### PR DESCRIPTION
**Pull request by tomaszszymanski129**
I want to merge this change because it allows to manipulate order lines for unconfirmed orders.

Consists of already reviewed changes.

- DraftOrderLine[CRUD] refactored to non-draft, changed it's place and allowed draft and unconfirmed orders
- Refactor old events which were draft (`order_added_products`, `order_removed_products`) - these two are now created for both draft and unconfirmed.
- Data migraton for events above (change existing DB names)
- Extra tests for events and order status

Dashboard frontend available on branch:
`feature/unconfirmed-order-line-manipulation`

Related frontend PR: https://github.com/mirumee/saleor-dashboard/pull/967

# Impact

* [x] New migrations
* [x] New/Updated API fields or mutations
* [ ] Deprecated API fields or mutations
* [x] Removed API types, fields, or mutations
* [ ] Documentation needs to be updated

# Pull Request Checklist

* [ ] Privileged queries and mutations are guarded by proper permission checks
* [ ] Database queries are optimized and the number of queries is constant
* [x] Database migration files are up to date
* [x] The changes are tested
* [x] GraphQL schema and type definitions are up to date
* [ ] Changes are mentioned in the changelog

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -3,6 +3,7 @@
 All notable, unreleased changes to this project will be documented in this file. For the released changes, please visit the [Releases](https://github.com/mirumee/saleor/releases) page.
 
 ## [Unreleased]
+- Unconfirmed order manipulation - #6829 by @tomaszszymanski129
```
Comment (IKarbowiak) [inline CHANGELOG.md:6]: You should put this under `Breaking` or `Other` section.

**Review by IKarbowiak — APPROVED** [review]

**Review by maarcingebala — APPROVED** [review]

**Comment by maarcingebala** [comment]
Currently admin can remove all products from the order. We should disallow confirming such orders by raising a validation error.

**Review by fowczarek — CHANGES_REQUESTED** [review]
Probably bug or invalid configuration:
1. Set avalara to auto-commit.
2. Creat unconfirmed order 
3. When you edit the order there are no changes  in AvaTax

**Comment by fowczarek** [comment]
Bug:
Order is updated in AvaTax only when adding a new product to the order. There is no AvaTax refresh when the order has updated quantity or when a line is deleted from the order.

**Comment by fowczarek** [comment]
Bug: (Invalid tax calculation)
1. Create any unconfirmed order.
2. Add any product 
3. Increase the quantity of added products to at least 10.
4. Add any other product 
5. Taxes for the product with incrised quantity are invalid.
image
image

**Comment by tomaszszymanski129** [comment]
Avalara should now behave correctly.
- Order is updated in avatax when products are added, removed, quantity changes
- Tax data is fetched properly and saved on model on each above action
- Orders `total` matches what we see in AvaTax

**Comment by fowczarek** [comment]
After add product to unconfirmed order `Taxes (VAT included)` is not updated. 
Affter add `Colored Parrot Cushion` tax in avalara is `10.93` and in Saleor we have `7.19` 
image
image

**Comment by fowczarek** [comment]
After shipping method change Tax is not updated. 
In Saleor: 7.19
In Avatax: 19.85
image
image

**Comment by fowczarek** [comment]
The quantity update doesn't work correctly. (After update total price is the same that before the update but the unit price is increased or decrised) 

Before quantity update we have:
2x Wine $5 each
1x Juice $7 

After the update we have:
1x Wine $10
10x Juice $0,7 each 

image
image

**Comment by fowczarek** [comment]
You have a similar issue to https://app.clickup.com/t/2549495/SALEOR-2262

When an order is in PLN-channel, prices displayed in "Add product" view are in USD but should be in PLN in my case.

File: saleor/graphql/order/mutations/orders.py [hunk 2]
```diff
@@ -543,3 +570,187 @@ def perform_mutation(cls, root, info, **data):
...
+        try:
+            channel = order.channel
+            validate_product_is_published_in_channel(variants, channel)
+            validate_variant_channel_listings(variants, channel)
+        except ValidationError as error:
+            raise ValidationError({"input": error})
+        # Add the lines
+        try:
+            lines = [
+                add_variant_to_order(order, variant, quantity)
```
[the remark this answers is not in the document]
Comment (maarcingebala) [reply saleor/graphql/order/mutations/orders.py:643]: There is `add_variants_to_checkout` and we should have a similar function for adding variants to order, but I would say it's not in the scope of this task.

File: saleor/graphql/schema.graphql [hunk 3]
```diff
@@ -3075,8 +3054,8 @@ enum OrderEventsEmailsEnum {
 enum OrderEventsEnum {
   DRAFT_CREATED
   DRAFT_CREATED_FROM_REPLACE
-  DRAFT_ADDED_PRODUCTS
-  DRAFT_REMOVED_PRODUCTS
+  ADDED_PRODUCTS
```
[the remark this answers is not in the document]
Comment (tomaszszymanski129) [reply saleor/graphql/schema.graphql:3057]: Yes we did 👍🏻

File: saleor/graphql/order/mutations/orders.py [hunk 4]
```diff
@@ -543,3 +570,187 @@ def perform_mutation(cls, root, info, **data):
...
+    class Meta:
+        description = "Create order lines for an order."
+        permissions = (OrderPermissions.MANAGE_ORDERS,)
+        error_type_class = OrderError
+        error_type_field = "order_errors"
+
+    @classmethod
+    def perform_mutation(cls, _root, info, **data):
+        order = cls.get_node_or_error(info, data.get("id"), only_type=Order)
+        if order.status not in ORDER_EDITABLE_STATUS:
```
[the remark this answers is not in the document]
Comment (tomaszszymanski129) [reply saleor/graphql/order/mutations/orders.py]: moved it all to separate methods, it's much cleaner now.

**Comment by fowczarek** [comment]
Changing shipping address doesn't trigger Avalara update. 

image
image

**Comment by tomaszszymanski129** [comment]
@fowczarek Avalara now triggers correctly and taxes for each item, shipping and total are re-calculated using new address.

File: saleor/order/utils.py [hunk 5]
```diff
@@ -135,6 +135,16 @@ def update_order_prices(order, discounts):
             )
 
             price = manager.calculate_order_line_unit(order, line, variant, product)
+            if price is not None:
```
Comment (korycins) [inline saleor/order/utils.py:138]: Honestly, I don't understand the concept of this logic. Can you explain it?
- price always will be  different than None. When we don't have tax plugin enabled we will use price for calculations.py.
- You should not operate on .amount values. you should just call unit_price * quantity.  With that case, .quantitze is not needed. 
- I would recommend to update to the newest master. It contains changes in this method, but again IMO this part of code is not needed.

File: saleor/order/utils.py [hunk 6]
```diff
@@ -290,42 +303,53 @@ def add_gift_card_to_order(order, gift_card, total_price_left):
...
 
-def change_order_line_quantity(user, line, old_quantity, new_quantity):
+def change_order_line_quantity(context, line, old_quantity, new_quantity):
     """Change the quantity of ordered items in a order line."""
     if new_quantity:
         line.quantity = new_quantity
-        total_price_net_amount = line.quantity * line.unit_price_net_amount
-        total_price_gross_amount = line.quantity * line.unit_price_gross_amount
+        line.save(update_fields=["quantity"])
+        net = line.unit_price.net.amount
```
Comment (korycins) [inline saleor/order/utils.py:311]: Why? what was wrong with this part of code? This logic was working well with the draft order so I don't understand why this changes are required.
Also unit_price returned from plugins will never be a None.

File: saleor/plugins/avatax/plugin.py [hunk 7]
```diff
@@ -119,15 +119,6 @@ def _skip_plugin(
         if not self.active:
             return True
 
-        # The previous plugin already calculated taxes so we can skip our logic
```
Comment (korycins) [inline saleor/plugins/avatax/plugin.py:122]: Why did you remove this part of code? 
```
if isinstance(previous_value, TaxedMoney):
    return previous_value.net != previous_value.gross
``` 
Allows us to not calculate taxes twice for the same line.

File: saleor/plugins/avatax/plugin.py [hunk 8]
```diff
@@ -452,6 +443,14 @@ def calculate_order_shipping(
             gross=order.shipping_method.price,  # type: ignore
         )
 
+    def order_line_deleted(
```
Comment (korycins) [inline saleor/plugins/avatax/plugin.py:446]: Again, this is not a needed part of the code.

**Review by korycins — CHANGES_REQUESTED** [review]

File: saleor/graphql/order/mutations/orders.py [hunk 9]
```diff
@@ -538,6 +561,15 @@ def get_instance(cls, info, **data):
                     )
                 }
             )
+        if not instance.lines.count():
```
Comment (maarcingebala) [inline saleor/graphql/order/mutations/orders.py:564]: `exists` should be faster when we only want to check the existence of an object.

```suggestion
        if not instance.lines.exists():
```

File: saleor/graphql/order/mutations/orders.py [hunk 10]
```diff
@@ -574,3 +607,219 @@ def perform_mutation(cls, root, info, **data):
...
+        )
+        order = line.order
+        cls.validate_order(line.order)
+
+        db_id = line.id
+        line_info = OrderLineData(
+            line=line,
+            quantity=line.quantity,
+            variant=line.variant,
+            warehouse_pk=line.allocations.first().stock.warehouse.pk,
```
Comment (fowczarek) [inline saleor/graphql/order/mutations/orders.py:754]: If you perform this mutation for draft order then the line doesn't have any allocation. We should fetch `warehouse_pk` only for unconfirmed orders.

File: saleor/graphql/order/mutations/orders.py [hunk 11]
```diff
@@ -574,3 +607,219 @@ def perform_mutation(cls, root, info, **data):
...
+            )
+        return cleaned_input
+
+    @classmethod
+    def save(cls, info, instance, cleaned_input):
+        line_info = OrderLineData(
+            line=instance,
+            quantity=instance.quantity,
+            variant=instance.variant,
+            warehouse_pk=instance.allocations.first().stock.warehouse.pk,
```
Comment (fowczarek) [inline saleor/graphql/order/mutations/orders.py:808]: If you perform this mutation for draft order then the line doesn't have any allocation. We should fetch `warehouse_pk` only for unconfirmed orders.

**Review by fowczarek — CHANGES_REQUESTED** [review]

**Review by fowczarek — APPROVED** [review]

File: saleor/graphql/schema.graphql [hunk 12]
```diff
@@ -2725,9 +2704,6 @@ type Mutation {
   draftOrderDelete(id: ID!): DraftOrderDelete
   draftOrderBulkDelete(ids: [ID]!): DraftOrderBulkDelete
   draftOrderLinesBulkDelete(ids: [ID]!): DraftOrderLinesBulkDelete
-  draftOrderLinesCreate(id: ID!, input: [OrderLineCreateInput]!): DraftOrderLinesCreate
```
Comment (korycins) [inline saleor/graphql/schema.graphql:2728]: The info about removing these mutations should be placed in Changelog in breakingChanes section. Someone could have an app which uses this.

File: saleor/order/__init__.py [hunk 13]
```diff
@@ -113,8 +113,8 @@ class OrderEvents:
     CHOICES = [
         (DRAFT_CREATED, "The draft order was created"),
         (DRAFT_CREATED_FROM_REPLACE, "The draft order with replace lines was created"),
-        (DRAFT_ADDED_PRODUCTS, "Some products were added to the draft order"),
-        (DRAFT_REMOVED_PRODUCTS, "Some products were removed from the draft order"),
+        (ADDED_PRODUCTS, "Some products were added to the order"),
```
Comment (korycins) [inline saleor/order/__init__.py:116]: IMHO This will raise an error (correct me if I am wrong). 
You're dropping the enums, which are stored in DB. Saleor will try to map the db events to graphql's OrderEvent which will raise an error as we don't have such event types anymore.

**Review by korycins — APPROVED** [review]
Looks good. Just to confirm, have you checked the new migration with the DB where, draft order events exist?

**Comment by tomaszszymanski129** [comment]
@korycins yep, checked both - migration applied / reversed with dashboard of feature branch / master.
